make creating initial vbox disks work with nixops 35ac0208

Uthar opened the pull request:
`_eval_flags` and `nix_exprs` are undefined in nixops 35ac0208, which causes an error when deploying with `targetEnv="virtualbox"` -  one can use `nixops.evaluation.eval` instead  to make it work.

full trace:

```
git............> creating VirtualBox VM...
git............> Virtual machine 'nixops-c9e5f533-02c9-11ec-8d2a-f01faf415644-git' is created and registered.
git............> UUID: 657a996c-25fd-4d9b-8403-c6b9b9e8559f
git............> Settings file: '/home/kpg/VirtualBox VMs/nixops-c9e5f533-02c9-11ec-8d2a-f01faf415644-git/nixops-c9e5f533-02c9-11ec-8d2a-f01faf415644-git.vbox'
git............> creating disk ‘disk1’...
Traceback (most recent call last):
  File "/nix/store/xj0midzl1ynx6s8w82x136xc96bg8qwl-python3.8-nixops-20210821/bin/.nixops-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/__main__.py", line 56, in main
    args.op(args)
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/script_defs.py", line 688, in op_deploy
    depl.deploy(
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1352, in deploy
    self.run_with_notify("deploy", lambda: self._deploy(**kwargs))
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1341, in run_with_notify
    f()
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1352, in <lambda>
    self.run_with_notify("deploy", lambda: self._deploy(**kwargs))
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1255, in _deploy
    nixops.parallel.run_tasks(
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/parallel.py", line 106, in run_tasks
    raise list(exceptions.values())[0]
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/parallel.py", line 70, in thread_fun
    work_result = (worker_fun(t), None, t.name)
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1207, in worker
    r.create(
  File "/nix/store/hd7zizr2v0fvi8sh733lkdvlb1b9gvsy-python3-3.8.9-env/lib/python3.8/site-packages/nixopsvbox/backends/virtualbox.py", line 397, in create
    + self.depl._eval_flags(self.depl.nix_exprs)
AttributeError: 'Deployment' object has no attribute '_eval_flags'

```

Uthar committed:
make it work with nixops 35ac0208

_eval_flags and nix_exprs are undefined in nixops 35ac0208

onny commented:
I'm facing the same issue while using `libvirt` as deployment backend :(
```
$ nixops deploy -d pi-nix-testing
database> Connecting to qemu:///system...
database> preparing disk image...
Traceback (most recent call last):
  File "/nix/store/m3g2f9fj7l0n8v3p42yyfpplyf0jibp1-python3.9-nixops-2.0.0/bin/.nixops-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/__main__.py", line 56, in main
    args.op(args)
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/script_defs.py", line 688, in op_deploy
    depl.deploy(
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/deployment.py", line 1352, in deploy
    self.run_with_notify("deploy", lambda: self._deploy(**kwargs))
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/deployment.py", line 1341, in run_with_notify
    f()
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/deployment.py", line 1352, in <lambda>
    self.run_with_notify("deploy", lambda: self._deploy(**kwargs))
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/deployment.py", line 1255, in _deploy
    nixops.parallel.run_tasks(
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/parallel.py", line 106, in run_tasks
    raise list(exceptions.values())[0]
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/parallel.py", line 70, in thread_fun
    work_result = (worker_fun(t), None, t.name)
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops/deployment.py", line 1207, in worker
    r.create(
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops_virtd/backends/libvirtd.py", line 211, in create
    self._prepare_storage_volume()
  File "/nix/store/3il5njybb5whz9lh9wk9k3c70kb5xcn9-python3-3.9.6-env/lib/python3.9/site-packages/nixops_virtd/backends/libvirtd.py", line 239, in _prepare_storage_volume
    + self.depl._eval_flags(self.depl.nix_exprs)
AttributeError: 'Deployment' object has no attribute '_eval_flags'
```

jbuchermn commented:
Why is this not merged? Still an issue.

CodeLongAndProsper90 commented:
Just ran into this issue today, it would be great if this could be merged

yajo commented:
The patch is incomplete. By not calling `nix-build`, it's not making sure the disk is created on the nix store. After merging this patch, `nixops deploy` still fails with:

```
machine-1.> VBoxManage: error: Could not find file for the medium '/nix/store/7xgrcfcmcbhwb1bjcs4nq4dr2lh3mi3n-virtualbox-nixops-22.05.vmdk' (VERR_FILE_NOT_FOUND)
machine-1.> VBoxManage: error: Details: code VBOX_E_FILE_ERROR (0x80bb0004), component MediumWrap, interface IMedium, callee nsISupports
machine-1.> VBoxManage: error: Context: "OpenMedium(Bstr(pszFilenameOrUuid).raw(), enmDevType, enmAccessMode, fForceNewUuidOnOpen, pMedium.asOutParam())" at line 187 of file VBoxManageDisk.cpp
Traceback (most recent call last):
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/parallel.py", line 70, in thread_fun
    work_result = (worker_fun(t), None, t.name)
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1207, in worker
    r.create(
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixopsvbox/backends/virtualbox.py", line 409, in create
    self._logged_exec(
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/backends/__init__.py", line 582, in _logged_exec
    return nixops.util.logged_exec(command, self.logger, **kwargs)
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/util.py", line 372, in logged_exec
    raise CommandFailed(err, res)
nixops.util.CommandFailed: command ‘['VBoxManage', 'clonehd', '/nix/store/7xgrcfcmcbhwb1bjcs4nq4dr2lh3mi3n-virtualbox-nixops-22.05.vmdk', '/var/home/yajo/VirtualBox VMs/nixops-94a4598f-fdde-11ec-968b-b025aa45f949-k3s-agent-1/disk1.vdi', '--format', 'VDI']’ failed on machine ‘k3s-agent-1’ (exit code 1)
```

Which basically means that `/nix/store/7xgrcfcmcbhwb1bjcs4nq4dr2lh3mi3n-virtualbox-nixops-22.05.vmdk` doesn't exist.

yajo committed:
fix: force building base image

Without this fix, the virtualbox disk may not exist when creating the machine.

The error is:

```
machine-1.> VBoxManage: error: Could not find file for the medium '/nix/store/7xgrcfcmcbhwb1bjcs4nq4dr2lh3mi3n-virtualbox-nixops-22.05.vmdk' (VERR_FILE_NOT_FOUND)
machine-1.> VBoxManage: error: Details: code VBOX_E_FILE_ERROR (0x80bb0004), component MediumWrap, interface IMedium, callee nsISupports
machine-1.> VBoxManage: error: Context: "OpenMedium(Bstr(pszFilenameOrUuid).raw(), enmDevType, enmAccessMode, fForceNewUuidOnOpen, pMedium.asOutParam())" at line 187 of file VBoxManageDisk.cpp
Traceback (most recent call last):
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/parallel.py", line 70, in thread_fun
    work_result = (worker_fun(t), None, t.name)
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/deployment.py", line 1207, in worker
    r.create(
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixopsvbox/backends/virtualbox.py", line 409, in create
    self._logged_exec(
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/backends/__init__.py", line 582, in _logged_exec
    return nixops.util.logged_exec(command, self.logger, **kwargs)
  File "/nix/store/ywh9crwf21xbhgw31fy5zpcc4z9p9gzb-python3-3.8.9-env/lib/python3.8/site-packages/nixops/util.py", line 372, in logged_exec
    raise CommandFailed(err, res)
nixops.util.CommandFailed: command ‘['VBoxManage', 'clonehd', '/nix/store/7xgrcfcmcbhwb1bjcs4nq4dr2lh3mi3n-virtualbox-nixops-22.05.vmdk', '/var/home/yajo/VirtualBox VMs/nixops-94a4598f-fdde-11ec-968b-b025aa45f949-k3s-agent-1/disk1.vdi', '--format', 'VDI']’ failed on machine ‘k3s-agent-1’ (exit code 1)
```

See https://github.com/nix-community/nixops-vbox/pull/28#issuecomment-1177446597.

For this to work, we need https://github.com/NixOS/nixops/pull/1529.

@moduon MT-904

yajo commented:
Please @Uthar merge https://github.com/Uthar/nixops-vbox/pull/1 so this PR is complete. It depends on https://github.com/NixOS/nixops/pull/1529 being merged and fixes the problem from https://github.com/nix-community/nixops-vbox/pull/28#issuecomment-1177446597.

Uthar committed:
Merge pull request #1 from moduon/work-with-nixops-master

fix: force building base image

Uthar commented:
Thank you @yajo - I merged https://github.com/Uthar/nixops-vbox/pull/1

lordcirth commented:
So, can this PR be merged now that dependencies are in place?

GeorgesAlkhouri commented:
Is there anything left to do here? Is help needed? I would love to try out nixops with virtualbox?

yajo commented:
AFAIK this just needs to be merged.

ngasull commented:
Any chance we got this merged @AmineChikhaoui ? I'm still stumbling on this in 2023, this fix would be more than welcome.

deepfire commented:
cc @Kranzes @adisbladis

AmineChikhaoui commented:
Sorry I didn't check nixops notifications that much since I haven't been using it for a while.
